tests/projects/test_projects.py

```diff
@@ -11,32 +14,95 @@
 from mlflow.utils.file_utils import TempDir
 from mlflow.utils import env
 
-from tests.projects.utils import TEST_PROJECT_DIR, GIT_PROJECT_URI
+from tests.projects.utils import TEST_PROJECT_DIR, GIT_PROJECT_URI, TEST_DIR
+
+
+def _assert_dirs_equal(expected, actual):
+    dir_comparison = filecmp.dircmp(actual, expected)
+    assert len(dir_comparison.left_only) == 0
+    assert len(dir_comparison.right_only) == 0
+    assert len(dir_comparison.diff_files) == 0
+    assert len(dir_comparison.funny_files) == 0
+
+
+def test_fetch_project(tmpdir):
+    # Creating a local Git repo containing a MLproject file.
+    local_git = tmpdir.join('git_repo').strpath
+    repo = git.Repo.init(local_git)
+    dir_util.copy_tree(src=TEST_PROJECT_DIR, dst=local_git)
+    repo.git.add(A=True)
+    repo.index.commit("test")
+    git_repo_uri = "file://" + os.path.abspath(local_git)
+
+    # Creating a local Git repo with a MLproject file in a subdirectory.
+    local_git_subdir = tmpdir.join('subdir_git_repo').strpath
+    repo = git.Repo.init(local_git_subdir)
+    dir_util.copy_tree(src=os.path.join(TEST_DIR, "resources"), dst=local_git_subdir)
+    repo.git.add(A=True)
+    repo.index.commit("test")
+    git_subdir_repo = "file://" + os.path.abspath(local_git_subdir)
+
+    # Test fetching a project to be run locally.
+    dst_dir = tmpdir.join('local').strpath
+
+    work_dir = mlflow.projects._fetch_project(uri=TEST_PROJECT_DIR, subdirectory='', version=None,
+                                              dst_dir=dst_dir, git_username=None,
+                                              git_password=None)[0]
+    _assert_dirs_equal(expected=TEST_PROJECT_DIR, actual=work_dir)
+
+    # Test fetching a project located in a Git repo subdirectory.
+    dst_dir = tmpdir.join('git-no-subdir').strpath
+    work_dir = mlflow.projects._fetch_project(uri=git_repo_uri, subdirectory='', version=None,
+                                              dst_dir=dst_dir, git_username=None,
+                                              git_password=None)[0]
+    dst_dir = tmpdir.join('git-subdir').strpath
+    work_dir2 = mlflow.projects._fetch_project(uri=git_subdir_repo, subdirectory='example_project',
+                                               version=None, dst_dir=dst_dir, git_username=None,
+                                               git_password=None)[0]
+    _assert_dirs_equal(expected=work_dir, actual=work_dir2)
```

Comment:
Should the expected value here be `local_git_subdir` to make the test more robust? i.e. if `_fetch_project` returns "/tmp" this test will pass as is. As a general takeaway - I'd recommend comparing the result of the method we're testing to an expected value that's not dependent on the method's correctness :)